Bugs fixes, Refactor, New widgets, Default URL values, URL value compression, Tests

pgarrett-scripps opened the pull request:
Been looking for a package to do this for awhile, so I'm very glad to have come across this. I'm using it in a lot of projects so figured I would give it a solid update. Sorry for the massive PR.

## Changes
- Improved code organization
- Added comprehensive type annotations
- Added support for `pills` and `segmented_control` widgets
- Added test suite for all supported widgets (except `pills` and `segmented_control` as their testing functions are not yet available)
- Implemented GitHub Actions CI/CD pipeline with tests running across multiple Python versions (3.9-3.12)
- Fixed multi-date input handling and setting date to None
- Widgets now preserve original types instead of casting everything to strings
- Automatic URL value generation on first widget initialization
- Widgets which support multiple states (slider, pills, etc.) can no longer be switched by editing query parameters
- Now uses Python's `inspect` module to ensure proper argument mapping
- Made `url_key` parameter optional when `key` is provided (if `url_key` is provided, it will be used for both)
- Added optional URL compression (useful for very long text_areas) by passing `compress=True` to widgets
- Added support for custom compression/encoding of URL values by passing appropriate functions to `compressor` and `decompressor` parameters
- Statefulness can be disabled by passing `stateful=False`

pgarrett-scripps committed:
Added new input widgets

pgarrett-scripps committed:
pytests

pgarrett-scripps committed:
updated examples with new widgets

pgarrett-scripps committed:
refactor with pyproject.toml

pgarrett-scripps committed:
Remove deleted files

pgarrett-scripps committed:
github action for pytest

pgarrett-scripps committed:
fix actions

pgarrett-scripps committed:
action

pgarrett-scripps committed:
action

pgarrett-scripps committed:
action

pgarrett-scripps committed:
update

pgarrett-scripps committed:
added handlers

pgarrett-scripps committed:
major refactor

pgarrett-scripps committed:
bump version, docs, bugs

pgarrett-scripps committed:
README

pgarrett-scripps committed:
docs

pgarrett-scripps committed:
added compression

pgarrett-scripps committed:
removed brotli dep

pgarrett-scripps committed:
more tests, update README

pgarrett-scripps committed:
CHANGES

pgarrett-scripps committed:
form bug fix, revert to stp

pgarrett-scripps committed:
new docs app

pgarrett-scripps committed:
linting + actions

pgarrett-scripps committed:
linting + actions

pgarrett-scripps commented:
This PR should still be backwards compatible with the previous versions of stp.

pgarrett-scripps committed:
compress form bug

pgarrett-scripps committed:
bug fix form compression

pgarrett-scripps committed:
started to add support for data_editor

pgarrett-scripps committed:
data_editor working

pgarrett-scripps committed:
version

pgarrett-scripps committed:
v1.2.0

pgarrett-scripps committed:
reqs

pgarrett-scripps commented:
Hey, any update on this? If you think it's a better idea, I can just make my PR a separate repo/package.

franekp commented:
@pgarrett-scripps  thank you very much for these improvements, I really appreciate the effort you put into this!

Sorry for late response - finally found some time to look through it.

The changes generally look good, so I merged as-is and uploaded to PyPI - your improvements are now available to everyone who uses the PyPI package.

One thing that I'd like to be changed back is the move from "url_key" to just "key". Both work, but you changed the official examples to use the latter.

I chose "url_key" over just "key" because it is explicit and you don't have to be familiar with streamlit-permalink to know what it means. From The Zen of Python: "explicit is better than implicit".

pgarrett-scripps commented:
@franekp thanks for taking the time to look over this PR. I'll go ahead an update the examples to use url_key again. I have another PR with some bug fixes, as well as support for the latest version of streamlit.  Anything else you would like to see before I send it your way?

Here are the new changes:
- Added support for `accept_new_options` to multiselect and selectbox
- added functions: `get_query_params` and `get_page_url`
- Added support for `stp.data_editor`
- Bug fix for widget url compression in forms

franekp commented:
>  I'll go ahead an update the examples to use url_key again.

Nice!

> I have another PR with some bug fixes, as well as support for the latest version of streamlit. Anything else you would like to see before I send it your way?

Thanks for asking! Go ahead and send the PR. Would be nice if you could test that the PyPI package works after your change from setup.py to pyproject.toml. I built and uploaded it but haven't had time to try installing and running it.
